python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py

```diff
@@ -0,0 +1,155 @@
+    group_name: Optional[str] = None,
+    resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+) -> Union[SourceAsset, "_SourceAsset"]:
+
+    if observe_fn is not None:
+        return _SourceAsset()(observe_fn)
+
+    return _SourceAsset(
+        name,
+        key_prefix,
+        metadata,
+        io_manager_key,
+        io_manager_def,
+        description,
+        partitions_def,
+        _metadata_entries,
+        group_name,
+        resource_defs,
+    )
+
+
+class _SourceAsset:
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        key_prefix: Optional[CoercibleToAssetKeyPrefix] = None,
+        metadata: Optional[MetadataUserInput] = None,
+        io_manager_key: Optional[str] = None,
+        io_manager_def: Optional[IOManagerDefinition] = None,
+        description: Optional[str] = None,
+        partitions_def: Optional[PartitionsDefinition] = None,
+        _metadata_entries: Optional[Sequence[Union[MetadataEntry, PartitionMetadataEntry]]] = None,
+        group_name: Optional[str] = None,
+        resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+    ):
+        self.name = name
+        if isinstance(key_prefix, str):
+            key_prefix = [key_prefix]
+        elif key_prefix is None:
+            key_prefix = []
+        self.key_prefix = key_prefix
+        self.metadata = metadata
+        self.io_manager_key = io_manager_key
+        self.io_manager_def = io_manager_def
+        self.description = description
+        self.partitions_def = partitions_def
+        self._metadata_entries = _metadata_entries
+        self.group_name = group_name
+        self.resource_defs = resource_defs
+
+    def __call__(self, observe_fn: SourceAssetObserveFunction) -> SourceAsset:
+        if observe_fn is not None:
+            source_asset_name = self.name or observe_fn.__name__
+            source_asset_key = AssetKey([*self.key_prefix, source_asset_name])
+
+            def fn(context: OpExecutionContext) -> None:
+                raw_observation = observe_fn(context)
+                metadata = _raw_observation_to_metadata(raw_observation)
+                context.log_event(
+                    AssetObservation(
+                        asset_key=source_asset_key,
+                        metadata=metadata,
+                    )
+                )
+
+            op = _Op(
+                name="__".join(source_asset_key.path).replace("-", "_"),
+                description=self.description,
+            )(fn)
+
+        return SourceAsset(
+            key=source_asset_key,
+            metadata=self.metadata,
+            io_manager_key=self.io_manager_key,
+            io_manager_def=self.io_manager_def,
+            description=self.description,
+            partitions_def=self.partitions_def,
+            _metadata_entries=self._metadata_entries,
+            group_name=self.group_name,
+            resource_defs=self.resource_defs,
+            node_def=op,
```

Comment:
I suspect that there are situations where it may be useful inside a sensor instead of inside a run.  E.g. perhaps the sensor produced by `build_asset_reconciliation_sensor` should invoke it as part of sensor execution.

1. Thoughts?
2. If so, then maybe the SourceAsset shouldn't have an op and should just have this function directly? Then we can convert it to an op later on.